### ISSUE TITLE
Support for Elastic Beanstalk's Periodic Tasks? (via cron.yml)

### ISSUE THREAD

**Issue by briantigerchow** [tawan/active-elastic-job#40]
Does this library support Elastic Beanstalk's [periodic tasks](http://docs.aws.amazon.com/elasticbeanstalk/latest/dg/using-features-managing-env-tiers.html#worker-periodictasks) via `cron.yml` file? I do not see documentation for such a feature.

**Comment by briantigerchow** [comment]
If not, here is my basic implementation that I am using now:

``` yaml
version: 1
cron:
      - name: 'HeartbeatJob'
        url: '/_jobs/heartbeat'
        schedule: "*/30 * * * * *"
      - name: 'FetchStreamMetricsJob'
        url: '/_jobs/fetch_stream_metrics'
        schedule: "*/5 * * * * *"
```

``` ruby
Rails.application.routes.draw do
  post '_jobs/:job', to: 'job_runner#perform_now'
end
```

``` ruby
class JobRunnerController < ApplicationController
  def perform_now

    # job_name e.g. 'fetch_stream_metrics' for # FetchStreamMetricsJob

    job_name = params[:job]
    job = "#{job_name}_job".titleize.gsub(' ', '').constantize.new
    job.perform_now
  end
end
```

**Comment by tawan** [comment]
Hi @briantigerchow,

Currently this gem does not support periodic tasks, but I have been evaluating this possibility of including a support for this feature. Thanks a lot for your submission and for sharing your implementation. I'll leave this issue open and I'll come back to it in the near future.

**Comment by karensg** [comment]
Good question. I am looking forward to hear your advice on how to set it up according to best practices.

**Comment by sauy7** [comment]
FWIW I've implemented the same as @briantigerchow (cron.yml) except that I wrap the `post '_jobs'` endpoint in a condition based on the Rails environment. My web tier is "production" and my worker tier is "production_worker". My routes file (simplified) then has an if condition:

```
if Rails.env.production_worker?
  post '_jobs' ...
else
  # regular web endpoints
end
```

Whilst setting up separate Rails environments may seem like overkill, I have other requirements such as my web app is configured to force SSL but the worker app responds on port 80.

**Comment by karensg** [comment]
I just implemented your proposal. Thanks

**Comment by masonjeffreys** [comment]
Instead of using the methods above, I added a check to the middleware.

If a request appears to be coming from SQS but doesn't originate from the ActiveElasticJob gem, we check to see if it matches the "_jobs" route. If so, it's considered to be from the cron.yml file and we perform the action immediate.

This allows me to leave config.force_ssl on even for cron jobs.

`module ActiveElasticJob
  module Rack
    # This middleware intercepts requests which are sent by the SQS daemon
    # running in {Amazon Elastic Beanstalk worker environments}[http://docs.aws.amazon.com/elasticbeanstalk/latest/dg/using-features-managing-env-tiers.html].
    # It does this by looking at the +User-Agent+ header.
    # Furthermore, it verifies the digest which is sent along with a legit SQS
    # message, and passed as an HTTP header in the resulting request.
    # The digest is based on Rails' +secrets.secret_key_base+.
    # Therefore, the application running in the web environment, which generates
    # the digest, and the application running in the worker
    # environment, which verifies the digest, have to use the *same*
    # +secrets.secret_key_base+ setting.
    class SqsMessageConsumer

      def call(env) #:nodoc:
        request = ActionDispatch::Request.new env
        if enabled? && aws_sqsd?(request)
          if originates_from_gem?(request)
            unless request.local? || sent_from_docker_host?(request)
              m = "Accepts only requests from localhost for job processing".freeze
              return ['403', {CONTENT_TYPE_HEADER_NAME => 'text/plain' }, [ m ]]
            end
            begin
              verify!(request)
              job = JSON.load(request.body)
              ActiveJob::Base.execute(job)
            rescue ActiveElasticJob::MessageVerifier::InvalidDigest => e
              return [
                '403',
                {CONTENT_TYPE_HEADER_NAME => 'text/plain' },
                ["incorrect digest"]]
            end
            return [
              OK_RESPONSE_CODE ,
              {CONTENT_TYPE_HEADER_NAME => CONTENT_TYPE },
              [ '' ]]
          elsif env["PATH_INFO"] == "/_jobs"
            ### if enabled and from aws_sqsd but did not originate from the gem, ###
            ### it may be a cron jobs from cron.yml. Perform that here ####
            begin
              job_name = request.headers['X-Aws-Sqsd-Taskname']
              job = job_name.constantize.new
              job.perform_now
               c = "Performed a cron job from sqs".freeze
              return [OK_RESPONSE_CODE, {CONTENT_TYPE_HEADER_NAME => 'text/plain' }, [ c ]]
            rescue
              return [
                '403',
                {CONTENT_TYPE_HEADER_NAME => 'text/plain' },
                ["Unsuccessful cron job attempt sqs for task: #{request.headers['X-Aws-Sqsd-Taskname']}"]]
            end
          end
        end
        @app.call(env)
      end

    end
  end
end`

**Comment by briantigerchow** [comment]
How does the new built-in support for this behavior work?

**Comment by tawan** [comment]
Hi all,

Support for periodic tasks has just been released in version v2.0.0. I also added documentation on its usage to the [README](https://github.com/tawan/active-elastic-job/blob/master/README.md#set-up-periodic-tasks-cron-jobs).
@briantigerchow
